src/query/api/v1/handler/placement/replace.go

```diff
@@ -0,0 +1,154 @@
+)
+
+const (
+	// ReplaceHTTPMethod is the HTTP method for the the replace endpoint.
+	ReplaceHTTPMethod = http.MethodPost
+
+	replacePathName = "replace"
+)
+
+var (
+	// M3DBReplaceURL is the url for the m3db replace handler (method POST).
+	M3DBReplaceURL = path.Join(handler.RoutePrefixV1, M3DBServicePlacementPathName, replacePathName)
+)
+
+// ReplaceHandler is the type for placement replaces.
+type ReplaceHandler Handler
+
+// NewReplaceHandler returns a new ReplaceHandler.
+func NewReplaceHandler(opts HandlerOptions) *ReplaceHandler {
+	return &ReplaceHandler{HandlerOptions: opts, nowFn: time.Now}
+}
+
+func (h *ReplaceHandler) ServeHTTP(serviceName string, w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	logger := logging.WithContext(ctx)
+
+	req, pErr := h.parseRequest(r)
+	if pErr != nil {
+		xhttp.Error(w, pErr.Inner(), pErr.Code())
+		return
+	}
+
+	placement, err := h.Replace(serviceName, r, req)
+	if err != nil {
+		status := http.StatusInternalServerError
+		if _, ok := err.(unsafeAddError); ok {
+			status = http.StatusBadRequest
+		}
+		logger.Error("unable to replace instance", zap.Any("error", err))
+		xhttp.Error(w, err, status)
+		return
+	}
+
+	placementProto, err := placement.Proto()
+	if err != nil {
+		logger.Error("unable to get placement protobuf", zap.Any("error", err))
+		xhttp.Error(w, err, http.StatusInternalServerError)
+		return
+	}
+
+	resp := &admin.PlacementGetResponse{
+		Placement: placementProto,
+		Version:   int32(placement.GetVersion()),
+	}
+
+	xhttp.WriteProtoMsgJSONResponse(w, resp, logger)
+}
+
+func (h *ReplaceHandler) parseRequest(r *http.Request) (*admin.PlacementReplaceRequest, *xhttp.ParseError) {
+	defer r.Body.Close()
+
+	req := &admin.PlacementReplaceRequest{}
+	if err := jsonpb.Unmarshal(r.Body, req); err != nil {
+		return nil, xhttp.NewParseError(err, http.StatusBadRequest)
+	}
+
+	return req, nil
+}
+
+// Replace replaces instances.
+func (h *ReplaceHandler) Replace(
+	serviceName string,
+	httpReq *http.Request,
+	req *admin.PlacementReplaceRequest,
+) (placement.Placement, error) {
+
+	candidates, err := ConvertInstancesProto(req.Candidates)
+	if err != nil {
+		return nil, err
+	}
+
+	serviceOpts := NewServiceOptions(serviceName, httpReq.Header, h.M3AggServiceOptions)
+	service, algo, err := ServiceWithAlgo(h.ClusterClient, serviceOpts, h.nowFn())
+	if err != nil {
+		return nil, err
+	}
+
+	if req.Force {
+		newPlacement, _, err := service.ReplaceInstances(req.LeavingInstanceIDs, candidates)
+		if err != nil {
+			return nil, err
+		}
+
+		return newPlacement, nil
+	}
+
+	curPlacement, version, err := service.Placement()
+	if err != nil {
+		return nil, err
+	}
+
+	// M3Coordinator isn't sharded, can't check if its shards are available.
+	if serviceName != M3CoordinatorServiceName {
+		if err := validateAllAvailable(curPlacement); err != nil {
+			return nil, err
+		}
+	}
+
+	newPlacement, err := algo.ReplaceInstances(curPlacement, req.LeavingInstanceIDs, candidates)
```

Comment:
can you add a comment about why you use the algo / check and set stuff

Reply:
Will do